pegnetMining/miner.go

```diff
@@ -0,0 +1,188 @@
+
+	// Factom blockchain related alerts
+	FactomMonitor common.IMonitor
+	OPRGrader     opr.IGrader
+
+	// miningContext can be passed around, if the context is canceled, everyone who owns this can
+	// be notified. If the context is cancelled, that means mining has stopped
+	contextMutex  sync.Mutex // Go race complaining about context access
+	miningContext context.Context
+	cancelMining  context.CancelFunc
+}
+
+func NewPegnetMiner(config *config.Config, monitor common.IMonitor, grader opr.IGrader) *PegnetMiner {
+	m := new(PegnetMiner)
+	m.FactomMonitor = monitor
+	m.OPRGrader = grader
+	m.ID = GetNextMinerID()
+	m.Config = config
+
+	return m
+}
+
+// StopMining will halt all mining on this miner. It can be restarted
+func (p *PegnetMiner) StopMining() bool {
+	p.contextMutex.Lock()
+	defer p.contextMutex.Unlock()
+	if p.cancelMining == nil {
+		return false // Not mining
+	}
+	// All contexts listening to this cancel will cancel when they
+	// hit.
+	p.cancelMining()
+	// Set the cancel to nil so we know we are not running.
+	// We already called cancel, so it's safe to lose it's reference
+	p.cancelMining = nil
+	return true
+}
+
+// IDString is purely for inner process tracking
+func (p *PegnetMiner) IDString() string {
+	return fmt.Sprintf("Miner-%d", p.ID)
+}
+
+// LaunchMiningThread starts a single mining goroutine that listens to the monitor.
+// Starts on minute 1 and writes data at minute 9
+func (p *PegnetMiner) LaunchMiningThread(verbose bool) {
+	mineLog := log.WithFields(log.Fields{"miner": p.ID})
+
+	// TODO: Also tell Factom Monitor we are done listening
+	alert := p.FactomMonitor.NewListener()
+	gAlert := p.OPRGrader.GetAlert(p.IDString())
+	// Tell OPR grader we are no longer listening
+	defer p.OPRGrader.StopAlert(p.IDString())
+
+	p.contextMutex.Lock()
+	p.miningContext, p.cancelMining = context.WithCancel(context.Background())
+	p.contextMutex.Unlock()
+
+	numMiners, _ := p.Config.Int("Miner.NumberOfMiners")
+	mining := false
+	var oprO *opr.OraclePriceRecord
+	var err error
+MiningLoop:
+	for {
+		var fds common.MonitorEvent
+		select {
+		case fds = <-alert:
+		case <-p.miningContext.Done(): // If cancelled
+			if oprO != nil {
+				close(oprO.StopMining)
+			}
+			return // Stop the mining
+		}
+
+		mineLog.WithFields(log.Fields{
+			"height": fds.Dbht,
+			"minute": fds.Minute,
+		}).Debug("Miner received alert")
+		switch fds.Minute {
+		case 1:
+			if !mining {
+				mining = true
+				oprO, err = opr.NewOpr(p.miningContext, p.ID, fds.Dbht, p.Config, gAlert)
+				if err == context.Canceled {
+					continue MiningLoop // OPR cancelled
+				}
+				if err != nil {
+					log.WithError(err).Fatal("Error creating an OPR.  Likely a config file issue")
+				}
+				go oprO.Mine(verbose)
+			}
+		case 9:
+			if mining {
+				close(oprO.StopMining)
+				mining = false
+				writeMiningRecord(oprO)
+				if verbose {
+					did := strings.Join(oprO.FactomDigitalID[:len(oprO.FactomDigitalID)-1], "-")
```

Comment:
not a fan of this. i think it's cleaner to just get rid of this at the source: https://github.com/pegnet/pegnet/blob/a39289989dcaadbac5639116e3a759b23f72877c/opr/opr.go#L292-L301

Reply:
Are you referring to the `did` thing?

Reply:
yes. you're splitting FactomDigitalID by "-" and removing the last element ("miner%d") and then replace the last element with your own miner id. I think it'd be cleaner to just remove the code that adds "miner%d" in the OPR creation

Reply:
I actually pulled this code from the old implementation (not saying it's correct, I'm just still getting up to speed on all the data structures).

That did line is only being used in a log statement. I'm still not sure exactly what you are suggesting.

Reply:
> That did line is only being used in a log statement. I'm still not sure exactly what you are suggesting.

oh i missed that part, i assumed it was doing something transformative. there's no change needed in this case, my apologies